packages/main/src/Menu.js

```diff
@@ -0,0 +1,489 @@
+			this._parentMenuItem = this._parentItemsStack.length ? this._parentItemsStack[this._parentItemsStack.length - 1] : undefined;
+		}
+	}
+
+	_prepareCurrentItems(items) {
+		this._currentItems = items.map((item, index) => {
+			return {
+				item,
+				position: index + 1,
+				ariaHasPopup: item.hasChildren ? "menu" : undefined,
+			};
+		});
+	}
+
+	_createSubMenu(item, openerId) {
+		const subMenu = document.createElement(this.constructor.getMetadata().getTag());
+
+		subMenu._isSubMenu = true;
+		subMenu.setAttribute("id", `submenu-${openerId}`);
+		subMenu._parentMenuItem = item;
+		const subItems = item.children;
+		let clonedItem,
+			idx;
+		for (idx = 0; idx < subItems.length; idx++) {
+			clonedItem = subItems[idx].cloneNode(true);
+			subMenu.appendChild(clonedItem);
+		}
+		this.staticAreaItem.shadowRoot.querySelector(".ui5-menu-submenus").appendChild(subMenu);
+		item._subMenu = subMenu;
+	}
+
+	_openItemSubMenu(item, opener, actionId) {
+		item._subMenu.showAt(opener);
+		item._preventSubMenuClose = true;
+		this._openedSubMenuItem = item;
+		this._subMenuOpenerId = actionId;
+	}
+
+	_closeItemSubMenu(item, forceClose) {
+		if (Object.keys(item).length) {
+			if (forceClose) {
+				item._preventSubMenuClose = false;
+				this._closeSubMenuPopover(item._subMenu, true);
+			} else {
+				setTimeout(() => this._closeSubMenuPopover(item._subMenu), 0);
+			}
+		}
+	}
+
+	_closeSubMenuPopover(subMenu, forceClose) {
+		if (subMenu && Object.keys(subMenu).length) {
+			const parentItem = subMenu._parentMenuItem;
+
+			if (forceClose || !parentItem._preventSubMenuClose) {
+				subMenu.close();
+				subMenu.remove();
+				parentItem._subMenu = {};
+				this._openedSubMenuItem = {};
+				this._subMenuOpenerId = "";
+			}
+		}
+	}
+
+	_prepareSubMenuDesktopTablet(item, opener, actionId) {
+		if (actionId !== this._subMenuOpenerId || item.hasChildren) {
+			// close opened sub-menu if there is any opened
+			this._closeItemSubMenu(this._openedSubMenuItem, true);
+		}
+		if (item.hasChildren) {
+			// create new sub-menu
+			this._createSubMenu(item, actionId);
+			this._openItemSubMenu(item, opener, actionId);
+		}
+		if (this._parentMenuItem) {
+			this._parentMenuItem._preventSubMenuClose = true;
+		}
+	}
+
+	_prepareSubMenuPhone(item, opener, actionId) {
+		this._currentItems = this._prepareCurrentItems(item.items);
+		this._parentMenuItem = item;
+		this._parentItemsStack.push(item);
+	}
+
+	_itemMouseOver(event) {
+		if (isDesktop()) {
+			// respect mouseover only on desktop
+			const opener = event.target;
+			const item = opener.associatedItem;
+			const hoverId = opener.getAttribute("id");
+
+			this._prepareSubMenuDesktopTablet(item, opener, hoverId);
+		}
+	}
+
+	_itemMouseOut(event) {
+		if (isDesktop()) {
+			// respect mouseout only on desktop
+			const item = event.target.associatedItem;
+			if (item.hasChildren && item._subMenu) {
+				// try to close the sub-menu
```

Comment:
Yepp! Let me explain: if the currently hovered submenu item also have submenu, it will be opened. On the other hand, if the currently hovered menu leaves it's hovered state (mouse goes out) - it should close. But, if mouse goes to the submenu just opened, the "parent"submenu should stay opened too. That's why there is a "try" to close the submenu, but if meanwhile its' own submenu is being hovered, the closing procedure will be cancelled.